### PR TITLE
samples: Add apps/pipes.jakt

### PR DESCRIPTION
**Pull request by Lenart12**
Simple project using inline cpp and system calls to
demonstrate pipe, fork and dup2.

### CODE REVIEW

**Review by linusg — CHANGES_REQUESTED** [review]
https://github.com/SerenityOS/jakt/blob/main/CONTRIBUTING.md#code-submission-policy

> Write the commit message subject line in the imperative mood ("Foo: Change the way dates work", not "Foo: Changed the way dates work").

File: samples/apps/pipes.jakt [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+    STDERR = 2
+}
+
+struct Pipe {
+    reading_end: c_int
+    writing_end: c_int
+}
+
+class System {
+    public function getErrno() -> c_int {
```
Comment (linusg) [inline samples/apps/pipes.jakt:14]: Jakt follows a similar coding style as serenity, and serenity uses `snake_case` for methods and omits `get_` from getters.

File: samples/apps/pipes.jakt [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+        unsafe { cpp {
+            "return errno;"
+        }}
```
Comment (linusg) [inline samples/apps/pipes.jakt:17]:
```suggestion
        unsafe {
            cpp {
                "return errno;"
            }
        }
```

File: samples/apps/pipes.jakt [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        // Unreachable code
+        return -1
+    }
+
+    public function pipe() throws -> Pipe  {
+        unsafe{
```
Comment (linusg) [inline samples/apps/pipes.jakt:23]: Not nearly all of this needs to be wrapped in unsafe. In fact, it looks like *only* the cpp block needs it.

File: samples/apps/pipes.jakt [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+                throw Error::from_errno(getErrno() as! i32)
+            }
+
+            // Unreachable code
+            return Pipe(reading_end: 0, writing_end: 1)
+        }
+    }
+
+    public function dup2(replace: c_int, with: c_int) throws -> c_int {
+        unsafe{
```
Comment (linusg) [inline samples/apps/pipes.jakt:42]: Likewise

File: samples/apps/pipes.jakt (hunk 3 above)
Comment (linusg) [reply samples/apps/pipes.jakt:23]: Also missing a space between the keyword and {

File: samples/apps/pipes.jakt [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+                throw Error::from_errno(getErrno() as! i32)
+            }
+
+            // Unreachable code
+            return -1
+        }
+    }
+
+    public function fork() throws -> c_int {
+        unsafe{
```
Comment (linusg) [inline samples/apps/pipes.jakt:61]: Likewise

File: samples/apps/pipes.jakt [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+            let mutable fork_ret : c_int = 0;
+
+            cpp {
+                "fork_ret = ::fork();"
+            }
+
+            if fork_ret as! i64 == -1 {
```
Comment (linusg) [inline samples/apps/pipes.jakt:68]: Do we have to cast this? `dup2_ret >= 0` works...

File: samples/apps/pipes.jakt [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+                return fork_ret
+            }
+
+            // Unreachable code
+            return -1
+        }
+    }
+
+    public function wait() throws -> c_int {
+        unsafe{
```
Comment (linusg) [inline samples/apps/pipes.jakt:80]: Likewise

Comment (linusg) [reply samples/apps/pipes.jakt:80]: I'll stop pointing these out now, please find the rest yourself :)

File: samples/apps/pipes.jakt [hunk 8]
```diff
@@ -0,0 +1,183 @@
+
```
Comment (linusg) [inline samples/apps/pipes.jakt:1]: Weird newline here

File: samples/apps/pipes.jakt [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+            let c_exit = exit_code as! c_int 
+            cpp {
+            "::_exit(c_exit);";
+            }
+        }
+    }
+}
+
+function main() {
+
```
Comment (linusg) [inline samples/apps/pipes.jakt:161]: Weird newline here

**Review by linusg — CHANGES_REQUESTED** [review]
Code style comments:

File: samples/apps/pipes.jakt (hunk 6 above)
Comment (Lenart12) [reply samples/apps/pipes.jakt:68]: Yes, for now (#465)

File: samples/apps/pipes.jakt (hunk 1 above)
Comment (Lenart12) [reply samples/apps/pipes.jakt:14]: Function name cant be just `errno` since its defined as a macro. Will rename to `syscall_error_number`

**Comment by Lenart12** [comment]
Changed commit message and implemented changes mentioned in the review. Some weird arithmetic which I don't like was needed because negated c_int needs to be fixed (to avoid things like `not (close_ret >= 0)`, `close_ret == -1` would be better!).

**Review by linusg — CHANGES_REQUESTED** [review]
`Adds` is not imperative mood

File: samples/apps/pipes.jakt [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+        }
+
+        // Unreachable code
+        return -1
+    }
+
+    public function read(fd: c_int, count: usize) throws -> String {
+        let mutable read_ret : c_int = 0;
+
+        unsafe {
```
Comment (linusg) [inline samples/apps/pipes.jakt:109]: not all of this needs to be in `unsafe`

**Comment by Lenart12** [comment]
Sorry this is taking so long, english is not my strong suit :(
Unsafe was done like that before because if I had two unsafe blocks the char buffer wouldnt be defined in the second block but I have refactored it so there is only one now.

**Pull request closed without merging**